PyTAPS/transports.py

```diff
@@ -0,0 +1,356 @@
+                self.at_eof = False
+    """ Function that blocks until new data has arrived
+    """
+    async def await_data(self):
+        waiter = self.loop.create_future()
+        self.waiters.append(waiter)
+        try:
+            await waiter
+        finally:
+            del self.waiters[0]
+
+    def send(self, data):
+        """ Function responsible for sending data. It decides which
+            protocol is used and then uses the appropriate functions
+        """
+        self.message_count += 1
+        if self.connection.state is not ConnectionState.ESTABLISHED:
+                print_time("SendError occured, connection is not established.",
+                           color)
+                if self.send_error:
+                    self.loop.create_task(self.send_error(message_count))
+                return
+        # Frame the data
+        if self.connection.framer:
+            self.loop.create_task(self.connection.framer.handle_new_sent_message(data, None, False))
+        else:
+            self.loop.create_task(self.write(data))
+        return self.message_count
+
+    async def write(self):
+        pass
+
+    def receive(self, min_incomplete_length, max_length):
+        if self.connection.framer:
+            self.loop.create_task(self.read_framed(min_incomplete_length,
+                                  max_length))
+        else:
+            self.loop.create_task(self.read(min_incomplete_length,
+                                  max_length))
+
+    async def read(self):
+        pass
+
+    async def close(self):
+        pass
+
+
+class UdpTransport(TransportLayer):
+
+    async def active_open(self, transport):
+        self.transport = transport
+        for t in self.connection.pending:
+            t.cancel()
+        print_time("Connected successfully UDP.", color)
+        self.connection.state = ConnectionState.ESTABLISHED
+        if self.connection.framer:
+            # Send a start even to the framer and wait for a reply
+            await self.connection.framer.handle_start(self.connection)
+        if self.connection.ready:
+            self.loop.create_task(self.connection.ready(self.connection))
+        return
+
+    async def write(self, data):
+        """ Sends udp data
+        """
+        print_time("Writing UDP data.", color)
+        #try:
+        # See if the udp flow was the result of passive or active open
+        if self.connection.active:
+            # Write the data
+            self.transport.sendto(data.encode())
+        else:
+            remote_address = self.remote_endpoint.address
+            remote_port = self.remote_endpoint.port
+            self.transport.sendto(data.encode(), (remote_address, remote_port))
+        """except:
+            print_time("SendError occured.", color)
+            if self.connection.send_error:
+                self.loop.create_task(self.connection.send_error(self.message_count))
+            return"""
```

Comment:
Why is this commented out? Has the error handling moved elsewhere? If so, we can probably remove this.

Reply:
Nope, forgot to uncomment it. Thanks for spotting it.